session/txn.go

```diff
@@ -428,7 +434,7 @@ func (st *TxnState) KeysNeedToLock() ([]kv.Key, error) {
 	}
 	keys := make([]kv.Key, 0, st.stmtBufLen())
 	if err := kv.WalkMemBuffer(st.stmtBuf, func(k kv.Key, v []byte) error {
-		if !keyNeedToLock(k, v) {
+		if !st.stmtBuf.IfKeyNeedLock(k) && !keyNeedToLock(k, v) {
```

Comment:
`stmtBuf` shared the same `needLockKey` map with `txnBuf`. After `RollbackStmt` the new keys added into `stmtBuf.needLockKey` will not be removed.

But it will not cause correctness issue:
* If the key still existed after retry, it will be locked;
* Otherwise, it will not be collected here.

Maybe it will lock more keys than need? But this way seems much simpler than implementing a rollbackable `needLockKey`.

Reply:
Is `MemBuffer.Discard` called when rollback? Seems we can move `needLockKey` into `Sandbox`, then we can rollback it.

Reply:
Yep. I have already done something simliar as a part of #18372.

https://github.com/pingcap/tidb/blob/761a961ba0fb2ff8d48deb8b46e417d3d4cbd18f/kv/memdb.go#L32-L71
https://github.com/pingcap/tidb/blob/761a961ba0fb2ff8d48deb8b46e417d3d4cbd18f/kv/sandbox.go#L109

Reply:
I'll try it following your implementation in the master.